### DISCUSSION TITLE
【打卡】当你感觉需要撰写注释时，请先尝试重构，试着让所有注释都变得多余。

### DISCUSSION THREAD

**Discussion by swlws** [swustzzh/coder_books#18] [General]
书名：重构 改善既有代码的设计第二版

章节：代码的坏味道

1. 何时开始重构？没有标准，没有任何量度规矩比得上见识广博者的直觉
2. 语义化命名。如果你想不出一个好名字，说明背后很可能潜藏着更深的设计问题。
3. 合并重复代码
4. 拆分过长的函数
   - 每当感觉需要以注释来说明点什么的时候，我们就把需要说明的东西写进一个独立函数中
5. 过长的参数列表
6. 用函数封装全局变量
7. 可变数据
   - 函数编程。如果要更新一个数据结构，就返回一份新的数据副本，旧的数据仍保持不变
8. 任何 switch 语句都应该用以多态取代条件表达式消除掉
9. 使用管道（filter、map）代替循环
10. 避免过多的设计。不要企图以各式各样的钩子和特殊情况来处理一些非必要的事情
11. 减短消息链。因为当链路中某一个点出问题时，会导致异常
12. 过大的类。机器能运行，但人不易阅读，必须拆分
13. 当你感觉需要撰写注释时，请先尝试重构，试着让所有注释都变得多余。

**Comment by swustzzh** [comment]
我是比较喜欢加注释，因为时间久了自己也会忘。13条讲的内容对于具体的业务来说感觉不是很适用，复杂业务的代码，还是加注释更好理解。

**Reply by swlws** [reply]
这个写的有歧义了，原文想表达的意思是下面这种场景：

```js
function fn(){
  // ... 多行code

  // 注释。。。。

  // ... 多行code

  // 注释。。。。

  // ... 多行code

}
```
想表达的意思是，在代码块中存在许多对代码的解释时，可以先考虑把这些代码抽离为`小函数`

**Reply by swustzzh** [reply]
'小函数'太多了会不会有问题？感觉函数的划分粒度也是个问题。

**Reply by swlws** [reply]
慧眼如炬，一眼就发现了小函数问题。

有一段关于小函数的阐述，大意是，一个工程中，最有活力、最有价值的就是这些小函数，通过小函数组合更有价值的业务。

另外，早期编码时代，不建议使用小函数，因为会带来很多资源损耗，但目前的硬件设施基础下，可以完全忽略这点代价。

此外，小函数多了，跳转也多。不过，不必过分担心这个问题，因为现在编辑器对跳转交互很好，一个Tab中，甚至可以看其它文件的代码
